src/components/implementation/no_interface/cFE_booter/osfilesys.c

```diff
@@ -1,280 +1,886 @@
+    return OS_FS_SUCCESS;
 }
 
-int32 open(const char *path) {
-    assert(fs[0].root && !fs[1].root);
-    struct fsobj *root = fs[0].root;
-    root = find(path, root);
-    if (!root) return OS_FS_ERROR;
-    root->refcnt++;
-    return root->filedes;
+/*
+ * currently returns name and in path field
+ * perhaps it would be a good idea to track path in fsobj
+ * but it looks like cFE uses this only to check is fd is valid
+ */
+int32 file_FDGetInfo(int32 FD, OS_FDTableEntry *fd_prop)
+{
+
+    int32 ret = chk_fd(FD);
+    if (ret != OS_FS_SUCCESS) return OS_FS_ERR_INVALID_FD;
+
+    struct fd *filedes = &fd_tbl[FD];
+    if (filedes->ino == 0) return OS_FS_ERR_INVALID_FD;
+
+    fd_prop->OSfd = FD;
+    memcpy(&fd_prop->Path, filedes->file->name, strlen(filedes->file->name));
+    fd_prop->User = 0;
+    fd_prop->IsValid = 1;
+    return OS_FS_SUCCESS;
 }
 
-int32 close(int32 filedes) {
-    uint32 index = filedes + 1;
-    if (files[index].filedes != filedes)   return OS_FS_ERROR;
-    if (files[index].refcnt < 1)           return OS_FS_ERROR;
-    files[index].refcnt--;
+
+/******************************************************************************
+** Dirent Level Methods
+******************************************************************************/
+
+// Internally, FDs are considered unused when ino == 0
+int32 newfd_get(int32 ino)
+{
+    uint32 count = 1;
+    struct fd *filedes;
+
+    while (count <= OS_MAX_NUM_OPEN_FILES + 1 && fd_tbl[count].ino != 0) {
+        count++;
+    }
+    if (count == OS_MAX_NUM_OPEN_FILES + 1) return OS_FS_ERROR;
+
+    filedes =  &fd_tbl[count];
+    filedes->ino = ino;
+    filedes->access = NONE;
+    return count;
+}
+
+int32 dir_open(char *path)
+{
+    struct fsobj *file;
+    file = file_find(path);
+    if (!file) return 0;
+    if (file->ino == 0) return 0;
+    if (file->type != FSOBJ_DIR) return 0;
+
+    int32 FD = newfd_get(file->ino);
+    if (FD > OS_MAX_NUM_OPEN_FILES + 1 || FD <= 0) return 0;
+    file->refcnt++;
+
+    struct fd *filedes = &fd_tbl[FD];
+    filedes->access = READ;
+    filedes->file = file;
+
+    struct dir_position *position = &filedes->position.dir_pos;
+    position->open_dir = file;
+    position->cur = file;
+    position->status = NORMAL;
+
+    return FD;
+}
+
+uint32 dir_close(int32 FD)
+{
+    if (FD > OS_MAX_NUM_OPEN_FILES + 1 || FD <= 0) return OS_FS_ERROR;
+    struct fd *filedes = &fd_tbl[FD];
+
+    filedes->ino = 0;
     return OS_FS_SUCCESS;
 }
 
-int32 read(int32  filedes, void *buffer, uint32 nbytes) {
-    if (!buffer)                                 return OS_FS_ERR_INVALID_POINTER;
-    if (files[filedes + 1].filedes != filedes)   return OS_FS_ERROR;
-    if (files[filedes + 1].refcnt < 1)           return OS_FS_ERROR;
-    if (nbytes == 0)                             return 0;
-    memcpy(buffer, files[filedes + 1].data, nbytes);
-    return nbytes;
+void dir_rewind(int32 FD)
+{
+    if (FD > OS_MAX_NUM_OPEN_FILES + 1 || FD <= 0) return;
```

Comment:
Rewind does not have anyway of failing.  its only guaranteed to rewind a file if it is passed a valid FD